core/state/database.go

```diff
@@ -129,20 +129,12 @@ type cachingDB struct {
 
 // OpenTrie opens the main account trie at a specific root hash.
 func (db *cachingDB) OpenTrie(root common.Hash) (Trie, error) {
-	tr, err := trie.NewSecure(root, db.db)
-	if err != nil {
-		return nil, err
-	}
-	return tr, nil
+	return trie.NewSecure(root, db.db)
 }
 
 // OpenStorageTrie opens the storage trie of an account.
 func (db *cachingDB) OpenStorageTrie(addrHash, root common.Hash) (Trie, error) {
-	tr, err := trie.NewSecure(root, db.db)
-	if err != nil {
-		return nil, err
-	}
-	return tr, nil
+	return trie.NewSecure(root, db.db)
```

Comment:
I think this is wrong, and there's a reason why it's written as it is. 
`Trie` is an interface, and the way you've written it now, it will return a `Trie` interface with `nil` value. Which is not the same as returning a `nil`. 
So anything that does

```
  if tr, _; OpenStorageTrie; tr != nil{
      tr.Foo()
  } 
```
will have nil dereference errors with this change. IIRC

Reply:
I believe you are correct. My assumption was that the error should always be checked and if there was a non-nil error, then the `Trie` interface with `nil` value would not be used. I made some additional changes to try and enforce that this was true and can do another pass to make sure.

If that invariant is enforced throughout the codebase, do you think this would be a good change or do you prefer the current invariant where the `Trie` interface value may still be checked if there is a non-nil error?

Reply:
I think it is good to keep the behaviour of `nil` being returned. I think there's been a couple of cases where this bit someone, and eventually we changed it to explicitly return `nil`, to avoid the situation where someone runs into it in the future. IMO it's a golang quirk which should be avoided.

Reply:
Fair enough. Closing this PR since it will not be getting merged.